json.go

```diff
@@ -0,0 +1,186 @@
+// the shortest possible encoding using integer values with sorted object keys.
+func CanonicalJSON(input []byte) ([]byte, error) {
+	sorted, err := SortJSON(input, make([]byte, 0, len(input)))
+	if err != nil {
+		return nil, err
+	}
+	return CompactJSON(sorted, make([]byte, 0, len(sorted))), nil
+}
+
+// SortJSON reencodes the JSON with the object keys sorted by lexicographically
+// by codepoint.
+func SortJSON(input, output []byte) ([]byte, error) {
+	// Skip to the first character that isn't whitespace.
+	var i int
+	firstChar := input[i]
+	for firstChar <= ' ' {
+		firstChar = input[i]
+		i++
+	}
+	if firstChar == '[' {
+		return sortJSONArray(input, output)
+	}
+	if firstChar == '{' {
+		return sortJSONObject(input, output)
+	}
+	return append(output, input...), nil
+}
+
+func sortJSONArray(input, output []byte) ([]byte, error) {
+	var err error
+	var array []json.RawMessage
+	if err = json.Unmarshal(input, &array); err != nil {
+		return nil, err
+	}
+	sep := byte('[')
+	for _, value := range array {
+		output = append(output, sep)
+		sep = ','
+		if output, err = SortJSON(value, output); err != nil {
+			return nil, err
+		}
+	}
+	if sep == '[' {
+		output = append(output, '[', ']')
+	} else {
+		output = append(output, ']')
+	}
+	return output, nil
+}
+
+func sortJSONObject(input, output []byte) ([]byte, error) {
+	var err error
+	var object map[string]json.RawMessage
+	if err = json.Unmarshal(input, &object); err != nil {
+		return nil, err
+	}
+	keys := make([]string, len(object))
+	var j int
+	for key := range object {
+		keys[j] = key
+		j++
+	}
+	sort.Strings(keys)
+	sep := byte('{')
+	for _, key := range keys {
+		output = append(output, sep)
+		sep = ','
+		var encoded []byte
+		if encoded, err = json.Marshal(key); err != nil {
+			return nil, err
+		}
+		output = append(output, encoded...)
+		output = append(output, ':')
+		if output, err = SortJSON(object[key], output); err != nil {
+			return nil, err
+		}
+	}
+	if sep == '{' {
+		output = append(output, '{', '}')
+	} else {
+		output = append(output, '}')
+	}
+	return output, nil
+}
+
+// CompactJSON makes the encoded JSON as small as possible by removing
+// whitespace and unneeded unicode escapes
+func CompactJSON(input, output []byte) []byte {
+	var i int
+	for i < len(input) {
+		c := input[i]
+		i++
+		if c <= ' ' {
+			continue
+		}
+		output = append(output, c)
+		if c == '"' {
+			for i < len(input) {
+				c = input[i]
+				i++
+				if c == '\\' {
+					escape := input[i]
+					i++
+					if escape == 'u' {
+						output, i = compactUnicodeEscape(input, output, i)
+					} else {
+						if escape == '/' {
+							output = append(output, escape)
+						} else {
+							output = append(output, c, escape)
+						}
+					}
+				} else {
+					output = append(output, c)
+				}
+				if c == '"' {
+					break
+				}
+			}
+		}
+	}
+	return output
+}
+
+func compactUnicodeEscape(input, output []byte, index int) ([]byte, int) {
```

Comment:
Not a single comment exists in this function, do you *really* think that's appropriate?